### PR TITLE
REF: Refactor reference handling to the BlockManager level for iterrows

### PR DESCRIPTION
**Pull request by phofl**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

let's see if this passes everything

### CODE REVIEW

File: pandas/core/internals/managers.py [hunk 1]
```diff
@@ -247,6 +247,16 @@ def _has_no_reference_block(self, blkno: int) -> bool:
         """
         return not self.blocks[blkno].refs.has_reference()
 
+    def add_references(self, mgr: BaseBlockManager) -> None:
+        """
+        Adds the references from one manager to another.
```
Comment (jorisvandenbossche) [inline pandas/core/internals/managers.py:252]: This assumes an identical block structure between both managers?

Comment (phofl) [reply pandas/core/internals/managers.py:252]: Yep, I’ll add a comment explaining this

Comment (phofl) [reply pandas/core/internals/managers.py:252]: Added

File: pandas/core/frame.py [hunk 2]
```diff
@@ -1396,10 +1396,9 @@ def iterrows(self) -> Iterable[tuple[Hashable, Series]]:
         for k, v in zip(self.index, self.values):
             s = klass(v, index=columns, name=k).__finalize__(self)
             if using_cow and self._mgr.is_single_block:
-                s._mgr.blocks[0].refs = self._mgr.blocks[0].refs  # type: ignore[union-attr]  # noqa
-                s._mgr.blocks[0].refs.add_reference(  # type: ignore[union-attr]
-                    s._mgr.blocks[0]  # type: ignore[arg-type, union-attr]
-                )
+                # Argument 1 to "add_references" of "BaseArrayManager" has incompatible
+                # type "Union[BlockManager, ArrayManager]"; expected "BaseArrayManager"
```
Comment (jorisvandenbossche) [inline pandas/core/frame.py:1400]: To be honest, this is kind of bullshit .. (and just noise to add as a comment) There is no easy way to let mypy understand such typing with a base class?

**Review by jorisvandenbossche — APPROVED** [review]

File: pandas/core/frame.py (hunk 2 above)
Comment (phofl) [reply pandas/core/frame.py]: Theoretically we could make mypy aware that using_copy_on_write() guards agains ArrayManager, but I am not that familiar with mypy :)

Removed.

**Comment by phofl** [comment]
Merging, this was only a removed comment

**Pull request merged**